File: src/main/java/net/axay/blueutils/sql/MariaDB.java
package net.axay.blueutils.sql;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.mariadb.jdbc.MariaDbPoolDataSource;

import java.io.Closeable;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;

public class MariaDB implements Closeable, AutoCloseable {

    private MariaDbPoolDataSource pool;

    private final SQLLoginInformation sqlLoginInformation;

    public MariaDB(@NotNull SQLLoginInformation sqlLoginInformation) {
        this.sqlLoginInformation = sqlLoginInformation;
    }

    /**
     * This method opens a new "connection" to the database.
     */
    public void connect() {

        try {

            pool = new MariaDbPoolDataSource();

            pool.setUser(sqlLoginInformation.getUser());
            pool.setServerName(sqlLoginInformation.getHost());
            pool.setPort(sqlLoginInformation.getPort());
            pool.setPassword(sqlLoginInformation.getPassword());
            pool.setDatabaseName(sqlLoginInformation.getDatabase());

            pool.setMaxPoolSize(32);
            pool.setMinPoolSize(8);

            System.out.println("Successfully set up connection pool to " + sqlLoginInformation.getHost() + ":" + sqlLoginInformation.getPort());

        } catch (SQLException e) {
            System.out.println("Failed to set up connection pool to " + sqlLoginInformation.getHost() + ":" + sqlLoginInformation.getPort());
            e.printStackTrace();
        }

    }

    public MariaDbPoolDataSource getPool() {
        return pool;
    }

    /**
     * Closes the connection pool.
     * This method should be invoked.
     */
    @Override
    public void close() {
        if (pool != null) pool.close();
    }

    /**
     * @param sql the sql statement which should be executed
     * @param parameters the parameters for the prepared statement
     *                   (in the correct order)
     * @return the result of the sql statement
     * or null if the pool is null
     * @throws SQLException if the statement could not be executed
     */
    @Nullable
    public ResultSet executePreparedStatement(String sql, Object... parameters) throws SQLException {

        if (pool == null) return null;

        try (
                Connection connection = getPool().getConnection();
                PreparedStatement preparedStatement = connection.prepareStatement(sql);
        ) {

            int i = 1;
            for (Object parameter : parameters) {
                try {
                    preparedStatement.setObject(i, parameter);
                } catch (Exception exc) {
                    throw new IllegalArgumentException("An object of the type " + parameter.getClass().getName() + " cannot be set as a parameter of the prepared statement!");
                }
                i++;
            }

            return preparedStatement.executeQuery();

        }

    }

    /**
     * @return This will return true if the connection is still open.
     * This method won't validate the connection.
     * For that see {@link #validate(Connection, int)}
     */
    public static boolean isConnected(Connection connection) {

        try {

            if (connection == null) {
                return false;
            } else {
                return !connection.isClosed();
            }

        } catch (SQLException e) {
            e.printStackTrace();
            return false;
        }

    }

    /**
     * This method will validate the connection with
     * the given timeout amount.
     * @param timeout the amount of time in seconds the connection
     *                has time to validate itself
     * @return This will return true if the connection is still open.
     */
    public static boolean validate(Connection connection, int timeout) {

        try {

            if (!isConnected(connection)) {
                return false;
            } else {
                return connection.isValid(timeout);
            }

        } catch (SQLException e) {
            e.printStackTrace();
            return false;
        }

    }

}

File: src/main/java/net/axay/blueutils/sql/MariaDB.kt
package net.axay.blueutils.sql

import org.mariadb.jdbc.MariaDbPoolDataSource
import java.sql.ResultSet
import java.sql.SQLException

class MariaDB(private val databaseLoginInformation: DatabaseLoginInformation) : AutoCloseable {

    private var pool: MariaDbPoolDataSource? = null

    fun connect() {
        try {

            MariaDbPoolDataSource().let {

                it.user = databaseLoginInformation.user
                it.serverName = databaseLoginInformation.host
                it.port = databaseLoginInformation.port
                it.setPassword(databaseLoginInformation.password)
                it.databaseName = databaseLoginInformation.database
                it.maxPoolSize = 32
                it.minPoolSize = 8

                pool = it

            }

            println("Successfully set up connection pool to " + databaseLoginInformation.host + ":" + databaseLoginInformation.port)

        } catch (e: SQLException) {
            println("Failed to set up connection pool to " + databaseLoginInformation.host + ":" + databaseLoginInformation.port)
            e.printStackTrace()
        }
    }

    /**
     * Closes the connection pool.
     * This method should be invoked.
     */
    override fun close() {
        pool?.close()
    }

    /**
     * @param sql the sql statement which should be executed
     * @param parameters the parameters for the prepared statement
     * (in the correct order)
     * @return the result of the sql statement
     * @throws SQLException if the statement could not be executed
     * @throws IllegalArgumentException if the type of a parameter
     * cannot be used in a sql statement
     * @throws IllegalStateException if the pool is null
     */
    fun executePreparedStatement(sql: String, vararg parameters: Any): ResultSet {

        pool?.let { pool ->

            pool.connection.use { connection ->
            connection.prepareStatement(sql).use { preparedStatement ->

                for ((index: Int, parameter: Any) in parameters.withIndex()) {
                    try {
                        preparedStatement.setObject(index + 1, parameter)
                    } catch (exc: Exception) {
                        throw IllegalArgumentException("An object of the type ${parameter::class.simpleName} cannot be set as a parameter of the prepared statement!")
                    }
                }

                return preparedStatement.executeQuery()

            } }

        }

        throw IllegalStateException("Cannot execute prepared statement. Call connect() first!")

    }

}